File: src/cavebot/navigation/pathfinder.py
"""A* pathfinder on the WaypointCost grid."""

from __future__ import annotations

import heapq
import math

from cavebot.navigation.atlas import Atlas

_DIRECTIONS = [
    (0, -1, 1.0),
    (0, 1, 1.0),
    (-1, 0, 1.0),
    (1, 0, 1.0),
    (-1, -1, 1.414),
    (1, -1, 1.414),
    (-1, 1, 1.414),
    (1, 1, 1.414),
]

_BLOCKED_COST = 255
_UNEXPLORED_COST = 0


def _octile_distance(x1: int, y1: int, x2: int, y2: int) -> float:
    dx = abs(x2 - x1)
    dy = abs(y2 - y1)
    return max(dx, dy) + (math.sqrt(2) - 1) * min(dx, dy)


def find_path(
    atlas: Atlas,
    start: tuple[int, int, int],
    goal: tuple[int, int, int],
    max_iterations: int = 100_000,
) -> list[tuple[int, int]] | None:
    sx, sy, sz = start
    gx, gy, gz = goal

    if sz != gz:
        return None

    z = sz

    start_cost = atlas.get_cost_at(sx, sy, z)
    goal_cost = atlas.get_cost_at(gx, gy, z)
    if start_cost >= _BLOCKED_COST or start_cost == _UNEXPLORED_COST:
        return None
    if goal_cost >= _BLOCKED_COST or goal_cost == _UNEXPLORED_COST:
        return None

    if sx == gx and sy == gy:
        return [(sx, sy)]

    counter = 0
    open_set: list[tuple[float, int, int, int]] = []
    heapq.heappush(open_set, (_octile_distance(sx, sy, gx, gy), counter, sx, sy))
    counter += 1

    came_from: dict[tuple[int, int], tuple[int, int]] = {}
    g_score: dict[tuple[int, int], float] = {(sx, sy): 0.0}

    iterations = 0
    while open_set and iterations < max_iterations:
        iterations += 1
        _, _, cx, cy = heapq.heappop(open_set)

        if cx == gx and cy == gy:
            path = [(cx, cy)]
            while (cx, cy) in came_from:
                cx, cy = came_from[(cx, cy)]
                path.append((cx, cy))
            path.reverse()
            return path

        for dx, dy, move_cost_mult in _DIRECTIONS:
            nx, ny = cx + dx, cy + dy
            tile_cost = atlas.get_cost_at(nx, ny, z)
            if tile_cost >= _BLOCKED_COST or tile_cost == _UNEXPLORED_COST:
                continue

            move_cost = move_cost_mult * (tile_cost / 150.0)
            tentative_g = g_score[(cx, cy)] + move_cost

            if tentative_g < g_score.get((nx, ny), float("inf")):
                came_from[(nx, ny)] = (cx, cy)
                g_score[(nx, ny)] = tentative_g
                f = tentative_g + _octile_distance(nx, ny, gx, gy)
                heapq.heappush(open_set, (f, counter, nx, ny))
                counter += 1

    return None

File: go-cavebot/internal/navigation/pathfinder.go
package navigation

import (
	"container/heap"
	"math"
)

const (
	blockedCost    = 255
	unexploredCost = 0
)

type movDir struct {
	dx, dy int
	cost   float64
}

var moveDirs = []movDir{
	{0, -1, 1.0},
	{0, 1, 1.0},
	{-1, 0, 1.0},
	{1, 0, 1.0},
	{-1, -1, 1.414},
	{1, -1, 1.414},
	{-1, 1, 1.414},
	{1, 1, 1.414},
}

func octileDistance(x1, y1, x2, y2 int) float64 {
	dx := math.Abs(float64(x2 - x1))
	dy := math.Abs(float64(y2 - y1))
	return math.Max(dx, dy) + (math.Sqrt2-1)*math.Min(dx, dy)
}

// FindPath runs A* from start to goal on the atlas cost grid.
// start and goal are [x, y, z]. Returns a slice of [x, y] points or nil.
func FindPath(atlas *Atlas, start, goal [3]int, maxIterations int) [][2]int {
	sx, sy, sz := start[0], start[1], start[2]
	gx, gy, gz := goal[0], goal[1], goal[2]

	if sz != gz {
		return nil
	}
	z := sz

	startCost := atlas.GetCostAt(sx, sy, z)
	goalCost := atlas.GetCostAt(gx, gy, z)
	if startCost >= blockedCost || startCost == unexploredCost {
		return nil
	}
	if goalCost >= blockedCost || goalCost == unexploredCost {
		return nil
	}

	if sx == gx && sy == gy {
		return [][2]int{{sx, sy}}
	}

	type point = [2]int
	counter := 0
	pq := &priorityQueue{}
	heap.Init(pq)
	heap.Push(pq, &pqItem{
		x: sx, y: sy,
		f:   octileDistance(sx, sy, gx, gy),
		idx: counter,
	})
	counter++

	cameFrom := make(map[point]point)
	gScore := make(map[point]float64)
	gScore[point{sx, sy}] = 0

	iterations := 0
	for pq.Len() > 0 && iterations < maxIterations {
		iterations++
		cur := heap.Pop(pq).(*pqItem)
		cx, cy := cur.x, cur.y

		if cx == gx && cy == gy {
			path := []point{{cx, cy}}
			for {
				prev, ok := cameFrom[point{cx, cy}]
				if !ok {
					break
				}
				cx, cy = prev[0], prev[1]
				path = append(path, point{cx, cy})
			}
			for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
				path[i], path[j] = path[j], path[i]
			}
			return path
		}

		for _, d := range moveDirs {
			nx, ny := cx+d.dx, cy+d.dy
			tileCost := atlas.GetCostAt(nx, ny, z)
			if tileCost >= blockedCost || tileCost == unexploredCost {
				continue
			}
			moveCost := d.cost * (float64(tileCost) / 150.0)
			tentativeG := gScore[point{cx, cy}] + moveCost

			prev, exists := gScore[point{nx, ny}]
			if !exists || tentativeG < prev {
				cameFrom[point{nx, ny}] = point{cx, cy}
				gScore[point{nx, ny}] = tentativeG
				f := tentativeG + octileDistance(nx, ny, gx, gy)
				heap.Push(pq, &pqItem{x: nx, y: ny, f: f, idx: counter})
				counter++
			}
		}
	}
	return nil
}

// Priority queue for A*
type pqItem struct {
	x, y    int
	f       float64
	idx     int
	heapIdx int
}

type priorityQueue []*pqItem

func (pq priorityQueue) Len() int            { return len(pq) }
func (pq priorityQueue) Less(i, j int) bool  { return pq[i].f < pq[j].f }
func (pq priorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].heapIdx = i
	pq[j].heapIdx = j
}
func (pq *priorityQueue) Push(x interface{}) {
	item := x.(*pqItem)
	item.heapIdx = len(*pq)
	*pq = append(*pq, item)
}
func (pq *priorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	old[n-1] = nil
	item.heapIdx = -1
	*pq = old[:n-1]
	return item
}
